plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java

```diff
@@ -321,32 +350,62 @@ public void testWriteLargeBlob() throws Exception {
         assertThat(countDownComplete.isCountedDown(), is(true));
     }
 
+    private static long checksum(final InputStream inputStream, final int length) throws IOException {
+        try (CRC32ChecksumCalculatingInputStream stream = new CRC32ChecksumCalculatingInputStream(inputStream)) {
+            BytesReference bytesReference = Streams.readFully(stream);
+            if (bytesReference.length() != length) {
+                throw new IllegalStateException("Not all bytes were read");
+            }
+            return stream.getCRC32Checksum();
+        }
+    }
+
     /**
      * A resettable InputStream that only serves zeros.
-     *
-     * Ideally it should be wrapped into a BufferedInputStream but it seems that the AWS SDK is calling InputStream{@link #reset()}
-     * before calling InputStream{@link #mark(int)}, which is not permitted by the {@link #reset()} method contract.
      **/
     private static class ZeroInputStream extends InputStream {
 
         private final AtomicBoolean closed = new AtomicBoolean(false);
+        private final byte[] buffer = new byte[1024];
         private final long length;
         private final AtomicLong reads;
         private volatile long mark;
 
         private ZeroInputStream(final long length) {
             this.length = length;
-            this.reads = new AtomicLong(length);
+            this.reads = new AtomicLong(0);
             this.mark = -1;
+            Arrays.fill(buffer, (byte) 0);
         }
 
         @Override
         public int read() throws IOException {
             ensureOpen();
-            if (reads.decrementAndGet() < 0) {
+            return (reads.incrementAndGet() <= length) ? 0 : -1;
+        }
+
+        @Override
+        public int read(byte[] b, int off, int len) throws IOException {
+            ensureOpen();
+            if (len == 0) {
+                return 0;
+            }
+            final int available = available();
+            if (available == 0) {
                 return -1;
             }
-            return 0;
+
+            final int toCopy = Math.min(len, available);
```

Comment:
Maybe I'm missing something here but can't we just do:

`Arrays.fill(b, off, off + len, (byte) 0);` and save ourselves the trouble of the buffer array and copying? :)

Reply:
I'm confused that I've not seen this myself :/

We still need to check that we are not reading more bytes than available in the stream. I pushed 96d3f140b3946e2524368ad3a4282a04474f0790